File: packages/sdk-typescript/src/trace.ts
import { randomUUID } from "node:crypto";
import { mkdirSync, writeFileSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import type {
  TraceEvent as CoreTraceEvent,
  TraceEventType,
  TraceStatus,
  Trace as CoreTrace,
  BudgetSummary,
} from "wickd-core";

export class TraceEvent {
  eventType: TraceEventType;
  timestamp: number;
  eventId: string;
  model?: string;
  provider?: string;
  inputTokens?: number;
  outputTokens?: number;
  cost?: number;
  latencyMs?: number;
  promptPreview?: string;
  responsePreview?: string;
  // Tool call (MCP or function call)
  toolName?: string;
  toolInputPreview?: string;
  toolOutputPreview?: string;
  toolStatus?: "success" | "error" | "denied";
  toolServer?: string;

  approvalName?: string;
  approvalStatus?: "approved" | "denied" | "pending";
  budgetTrigger?: string;
  spendAtEvent?: number;
  errorMessage?: string;
  metadata?: Record<string, unknown>;

  constructor(eventType: TraceEventType, fields: Partial<Omit<TraceEvent, "eventType">> = {}) {
    this.eventType = eventType;
    this.timestamp = fields.timestamp ?? Date.now() / 1000;
    this.eventId = fields.eventId ?? randomUUID().slice(0, 8);
    Object.assign(this, fields);
  }

  toDict(): CoreTraceEvent {
    const result: CoreTraceEvent = {
      eventType: this.eventType,
      timestamp: this.timestamp,
      eventId: this.eventId,
    };
    if (this.model != null) result.model = this.model;
    if (this.provider != null) result.provider = this.provider;
    if (this.inputTokens != null) result.inputTokens = this.inputTokens;
    if (this.outputTokens != null) result.outputTokens = this.outputTokens;
    if (this.cost != null) result.cost = this.cost;
    if (this.latencyMs != null) result.latencyMs = this.latencyMs;
    if (this.promptPreview != null) result.promptPreview = this.promptPreview;
    if (this.responsePreview != null) result.responsePreview = this.responsePreview;
    if (this.toolName != null) (result as unknown as Record<string, unknown>).toolName = this.toolName;
    if (this.toolInputPreview != null) (result as unknown as Record<string, unknown>).toolInputPreview = this.toolInputPreview;
    if (this.toolOutputPreview != null) (result as unknown as Record<string, unknown>).toolOutputPreview = this.toolOutputPreview;
    if (this.toolStatus != null) (result as unknown as Record<string, unknown>).toolStatus = this.toolStatus;
    if (this.toolServer != null) (result as unknown as Record<string, unknown>).toolServer = this.toolServer;
    if (this.approvalName != null) result.approvalName = this.approvalName;
    if (this.approvalStatus != null) result.approvalStatus = this.approvalStatus;
    if (this.budgetTrigger != null) result.budgetTrigger = this.budgetTrigger;
    if (this.spendAtEvent != null) result.spendAtEvent = this.spendAtEvent;
    if (this.errorMessage != null) result.errorMessage = this.errorMessage;
    if (this.metadata != null) result.metadata = this.metadata;
    return result;
  }
}

export class Trace {
  traceId: string;
  agentName: string;
  task: string;
  startedAt: number;
  endedAt?: number;
  events: TraceEvent[] = [];
  totalCost = 0;
  totalInputTokens = 0;
  totalOutputTokens = 0;
  status: TraceStatus = "running";
  budgetSummary?: BudgetSummary;

  constructor(agentName = "", task = "") {
    this.traceId = randomUUID();
    this.agentName = agentName;
    this.task = task;
    this.startedAt = Date.now() / 1000;
  }

  addEvent(event: TraceEvent): void {
    this.events.push(event);
    if (event.cost != null) this.totalCost += event.cost;
    if (event.inputTokens != null) this.totalInputTokens += event.inputTokens;
    if (event.outputTokens != null) this.totalOutputTokens += event.outputTokens;
  }

  addLlmCall(fields: {
    model: string;
    provider: string;
    inputTokens: number;
    outputTokens: number;
    cost: number;
    latencyMs: number;
    promptPreview?: string;
    responsePreview?: string;
  }): TraceEvent {
    const event = new TraceEvent("llm_call", {
      model: fields.model,
      provider: fields.provider,
      inputTokens: fields.inputTokens,
      outputTokens: fields.outputTokens,
      cost: fields.cost,
      latencyMs: fields.latencyMs,
      promptPreview: fields.promptPreview?.slice(0, 200),
      responsePreview: fields.responsePreview?.slice(0, 200),
      spendAtEvent: this.totalCost + fields.cost,
    });
    this.addEvent(event);
    return event;
  }

  addApproval(name: string, status: "approved" | "denied" | "pending"): TraceEvent {
    const event = new TraceEvent("approval_gate", {
      approvalName: name,
      approvalStatus: status,
      spendAtEvent: this.totalCost,
    });
    this.addEvent(event);
    return event;
  }

  addToolCall(fields: {
    toolName: string;
    latencyMs: number;
    toolInput?: string;
    toolOutput?: string;
    toolStatus?: "success" | "error" | "denied";
    toolServer?: string;
    metadata?: Record<string, unknown>;
  }): TraceEvent {
    const meta: Record<string, unknown> = { tool: fields.toolName, ...fields.metadata };
    if (fields.toolServer) meta.server = fields.toolServer;
    const event = new TraceEvent("tool_call", {
      toolName: fields.toolName,
      toolInputPreview: fields.toolInput?.slice(0, 200),
      toolOutputPreview: fields.toolOutput?.slice(0, 200),
      toolStatus: fields.toolStatus ?? "success",
      toolServer: fields.toolServer,
      latencyMs: fields.latencyMs,
      spendAtEvent: this.totalCost,
      metadata: meta,
    });
    this.addEvent(event);
    return event;
  }

  addBudgetKill(trigger: string, spend: number): TraceEvent {
    const event = new TraceEvent("budget_kill", {
      budgetTrigger: trigger,
      spendAtEvent: spend,
    });
    this.addEvent(event);
    this.status = "budget_killed";
    return event;
  }

  complete(budgetSummary?: BudgetSummary): void {
    this.endedAt = Date.now() / 1000;
    if (this.status === "running") this.status = "completed";
    this.budgetSummary = budgetSummary;
  }

  durationMs(): number | undefined {
    if (!this.endedAt) return undefined;
    return Math.round((this.endedAt - this.startedAt) * 1000 * 10) / 10;
  }

  toDict(): CoreTrace {
    return {
      traceId: this.traceId,
      agentName: this.agentName,
      task: this.task,
      startedAt: new Date(this.startedAt * 1000).toISOString(),
      endedAt: this.endedAt ? new Date(this.endedAt * 1000).toISOString() : undefined,
      durationMs: this.durationMs(),
      status: this.status,
      totalCost: Math.round(this.totalCost * 1_000_000) / 1_000_000,
      totalInputTokens: this.totalInputTokens,
      totalOutputTokens: this.totalOutputTokens,
      llmCalls: this.events.filter(e => e.eventType === "llm_call").length,
      toolCalls: this.events.filter(e => e.eventType === "tool_call").length,
      events: this.events.map(e => e.toDict()),
      budgetSummary: this.budgetSummary,
    };
  }

  toJSON(indent = 2): string {
    return JSON.stringify(this.toDict(), null, indent);
  }
}

export class TraceStore {
  readonly traceDir: string;

  constructor(traceDir?: string) {
    this.traceDir = traceDir ?? join(homedir(), ".wickd", "traces");
    mkdirSync(this.traceDir, { recursive: true });
  }

  save(trace: Trace): string {
    const filename = `${trace.agentName}_${trace.traceId.slice(0, 8)}.json`;
    const filepath = join(this.traceDir, filename);
    writeFileSync(filepath, trace.toJSON());
    return filepath;
  }

  listTraces(limit = 20): CoreTrace[] {
    try {
      const files = readdirSync(this.traceDir)
        .filter(f => f.endsWith(".json"))
        .map(f => ({ name: f, mtime: statSync(join(this.traceDir, f)).mtimeMs }))
        .sort((a, b) => b.mtime - a.mtime)
        .slice(0, limit);

      return files.map(({ name }) => {
        const raw = readFileSync(join(this.traceDir, name), "utf-8");
        return JSON.parse(raw);
      });
    } catch {
      return [];
    }
  }

  getTrace(traceId: string): CoreTrace | null {
    const files = readdirSync(this.traceDir);
    const match = files.find(f => f.includes(traceId));
    if (!match) return null;
    const raw = readFileSync(join(this.traceDir, match), "utf-8");
    return JSON.parse(raw);
  }
}

export class CloudTraceSync {
  private readonly endpoint: string;
  private readonly apiKey: string;

  constructor(endpoint: string, apiKey: string) {
    this.endpoint = endpoint.replace(/\/+$/, "");
    this.apiKey = apiKey;
  }

  /**
   * Fire-and-forget sync: sends trace to the cloud endpoint.
   * Does not block the agent. Errors are silently caught.
   */
  sync(trace: Trace): void {
    const body = JSON.stringify(trace.toDict());
    fetch(this.endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body,
    }).catch(() => {
      // Silently swallow errors — cloud sync is best-effort
    });
  }
}

File: packages/sdk-python/wickd/trace.py
import json
import logging
import os
import time
import uuid
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

logger = logging.getLogger("wickd")

MAX_PREVIEW = 200


@dataclass
class TraceEvent:
    """A single event in an agent trace.

    Fields are grouped by event type. Only populated fields are serialized.
    """

    event_type: str  # llm_call | tool_call | approval_gate | budget_kill | error
    timestamp: float = field(default_factory=time.time)
    event_id: str = field(default_factory=lambda: str(uuid.uuid4())[:8])

    # LLM call
    model: Optional[str] = None
    provider: Optional[str] = None
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    cost: Optional[float] = None
    latency_ms: Optional[float] = None
    prompt_preview: Optional[str] = None
    response_preview: Optional[str] = None

    # Tool call (MCP or function call)
    tool_name: Optional[str] = None
    tool_input_preview: Optional[str] = None
    tool_output_preview: Optional[str] = None
    tool_status: Optional[str] = None  # "success", "error", "denied"
    tool_server: Optional[str] = None  # MCP server name/origin

    # Approval
    approval_name: Optional[str] = None
    approval_status: Optional[str] = None

    # Budget
    budget_trigger: Optional[str] = None
    spend_at_event: Optional[float] = None

    # Error
    error_message: Optional[str] = None

    metadata: Optional[dict] = None

    def to_dict(self) -> dict:
        return {k: v for k, v in asdict(self).items() if v is not None}


@dataclass
class Trace:
    """A complete trace of one agent run."""

    trace_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    agent_name: str = ""
    task: str = ""
    started_at: float = field(default_factory=time.time)
    ended_at: Optional[float] = None
    events: list[TraceEvent] = field(default_factory=list)
    total_cost: float = 0.0
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    status: str = "running"
    budget_summary: Optional[dict] = None

    def add_event(self, event: TraceEvent):
        self.events.append(event)
        if event.cost:
            self.total_cost += event.cost
        if event.input_tokens:
            self.total_input_tokens += event.input_tokens
        if event.output_tokens:
            self.total_output_tokens += event.output_tokens

    def add_llm_call(
        self,
        model: str,
        provider: str,
        input_tokens: int,
        output_tokens: int,
        cost: float,
        latency_ms: float,
        prompt_preview: str = "",
        response_preview: str = "",
    ) -> TraceEvent:
        event = TraceEvent(
            event_type="llm_call",
            model=model,
            provider=provider,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cost=cost,
            latency_ms=latency_ms,
            prompt_preview=prompt_preview[:MAX_PREVIEW] if prompt_preview else None,
            response_preview=response_preview[:MAX_PREVIEW] if response_preview else None,
            spend_at_event=self.total_cost + cost,
        )
        self.add_event(event)
        return event

    def add_approval(self, name: str, status: str) -> TraceEvent:
        event = TraceEvent(
            event_type="approval_gate",
            approval_name=name,
            approval_status=status,
            spend_at_event=self.total_cost,
        )
        self.add_event(event)
        return event

    def add_tool_call(
        self,
        tool_name: str,
        latency_ms: float,
        tool_input: str = "",
        tool_output: str = "",
        tool_status: str = "success",
        tool_server: Optional[str] = None,
        metadata: Optional[dict] = None,
    ) -> TraceEvent:
        """Record an MCP or function tool call in the trace."""
        meta = {"tool": tool_name}
        if tool_server:
            meta["server"] = tool_server
        if metadata:
            meta.update(metadata)
        event = TraceEvent(
            event_type="tool_call",
            tool_name=tool_name,
            tool_input_preview=tool_input[:MAX_PREVIEW] if tool_input else None,
            tool_output_preview=tool_output[:MAX_PREVIEW] if tool_output else None,
            tool_status=tool_status,
            tool_server=tool_server,
            latency_ms=latency_ms,
            spend_at_event=self.total_cost,
            metadata=meta,
        )
        self.add_event(event)
        return event

    def add_budget_kill(self, trigger: str, spend: float) -> TraceEvent:
        event = TraceEvent(
            event_type="budget_kill",
            budget_trigger=trigger,
            spend_at_event=spend,
        )
        self.add_event(event)
        self.status = "budget_killed"
        return event

    def complete(self, budget_summary: Optional[dict] = None):
        self.ended_at = time.time()
        if self.status == "running":
            self.status = "completed"
        self.budget_summary = budget_summary

    def duration_ms(self) -> Optional[float]:
        if self.ended_at:
            return round((self.ended_at - self.started_at) * 1000, 1)
        return None

    def to_dict(self) -> dict:
        return {
            "trace_id": self.trace_id,
            "agent_name": self.agent_name,
            "task": self.task,
            "started_at": datetime.fromtimestamp(self.started_at, tz=timezone.utc).isoformat(),
            "ended_at": datetime.fromtimestamp(self.ended_at, tz=timezone.utc).isoformat() if self.ended_at else None,
            "duration_ms": self.duration_ms(),
            "status": self.status,
            "total_cost": round(self.total_cost, 6),
            "total_input_tokens": self.total_input_tokens,
            "total_output_tokens": self.total_output_tokens,
            "llm_calls": sum(1 for e in self.events if e.event_type == "llm_call"),
            "tool_calls": sum(1 for e in self.events if e.event_type == "tool_call"),
            "events": [e.to_dict() for e in self.events],
            "budget_summary": self.budget_summary,
        }

    def to_json(self, indent: int = 2) -> str:
        return json.dumps(self.to_dict(), indent=indent)


class TraceStore:
    """Saves traces as JSON to ~/.wickd/traces/."""

    def __init__(self, trace_dir: Optional[str] = None):
        self.trace_dir = Path(trace_dir) if trace_dir else Path.home() / ".wickd" / "traces"
        self.trace_dir.mkdir(parents=True, exist_ok=True)

    def save(self, trace: Trace) -> Path:
        filepath = self.trace_dir / f"{trace.agent_name}_{trace.trace_id[:8]}.json"
        filepath.write_text(trace.to_json())
        return filepath

    def list_traces(self, limit: int = 20) -> list[dict]:
        files = sorted(self.trace_dir.glob("*.json"), key=os.path.getmtime, reverse=True)
        traces = []
        for f in files[:limit]:
            try:
                traces.append(json.loads(f.read_text()))
            except (json.JSONDecodeError, OSError):
                continue
        return traces

    def get_trace(self, trace_id: str) -> Optional[dict]:
        for f in self.trace_dir.glob("*.json"):
            if trace_id in f.name:
                try:
                    return json.loads(f.read_text())
                except (json.JSONDecodeError, OSError):
                    return None
        return None


